logs_intake/socket/SendLogEntriesOverSocket.py

```diff
@@ -0,0 +1,104 @@
+# This code is provided as an example only.
+import json
+import time
+import socket
+import ssl
+
+ddApiKey = "<your_api_key>"
+
+# metadata: Additional metadata to send with the logs
+metadata = {
+    "ddsourcecategory": "aws",
+}
+
+host = "intake.logs.datadoghq.com"
+ssl_port = 10516
+
+DD_SOURCE = "ddsource"
+DD_CUSTOM_TAGS = "ddtags"
+
+metadata[DD_SOURCE] = "apigee"
+metadata[DD_CUSTOM_TAGS] = "level:info,x.api:identity-v1,env:dev,project:apigee"
+
+
+def lambda_handler(event):
+    # Check prerequisites
+    if ddApiKey == "<your_api_key>" or ddApiKey == "":
+        raise Exception(
+            "You must configure your API key before starting this lambda function (see #Parameters section)"
+        )
+
+    # Attach Datadog's Socket
+    s = connect_to_datadog(host, ssl_port)
+
+    try:
+        s = safe_submit_log(s, event)
+    except Exception as e:
+        print('Unexpected exception: {} for event {}'.format(str(e), event))
+    finally:
+        s.close()
+
+def connect_to_datadog(host, port):
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s = ssl.wrap_socket(s)
+    s.connect((host, port))
+    return s
+
+def safe_submit_log(s, log):
+    try:
+        send_entry(s, log)
+    except Exception as e:
+        # retry once
+        s = connect_to_datadog(host, ssl_port)
+        send_entry(s, log)
+    return s
+
+# Utility functions
+def send_entry(s, log_entry):
+    # The log_entry can only be a string or a dict
+    if isinstance(log_entry, str):
+        log_entry = {"message": log_entry}
+    elif not isinstance(log_entry, dict):
+        raise Exception(
+            "Cannot send the entry as it must be either a string or a dict. Provided entry: " + str(log_entry)
+        )
+    # Merge with metadata
+    log_entry = merge_dicts(log_entry, metadata)
+    print("log_entry has value {}".format(log_entry))
+
+    # Send to Datadog
+    str_entry = json.dumps(log_entry)
+    print(str_entry)
+    prefix = "%s " % ddApiKey
+    return s.send((prefix + str_entry + "\n").encode("UTF-8"))
+
+def merge_dicts(a, b, path=None):
+    if path is None:
+        path = []
+    for key in b:
+        if key in a:
+            if isinstance(a[key], dict) and isinstance(b[key], dict):
+                merge_dicts(a[key], b[key], path + [str(key)])
+            elif a[key] == b[key]:
+                pass  # same leaf value
+            else:
+                raise Exception(
+                    'Conflict while merging metadatas and the log entry at %s' % '.'.join(path + [str(key)])
+                )
+        else:
+            a[key] = b[key]
+    return a
+
+if __name__ == "__main__":
+    # execute only if run as a script
+    metadata[DD_CUSTOM_TAGS] = "x.error,level:error,x.api-identity-v1,x.api:identity-v1,x.env-prd,x.env:prd,x.proj-apigee,x.proj:apigee"
```

Comment:
what are all these? are these supposed to be examples of the script in action? can we please dumb it down to make it easier for the reader to modify to fit their needs?

Reply:
Cleaned-up somewhat.  Reduced down to 2 examples.

Reply:
Made examples simpler